packages/opentelemetry-core/src/trace/TraceState.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@opentelemetry/types';
+
+// TODO validate maximum number of items
+// const maximumTraceStateItems = 32;
+
+const maximumTraceStateLength = 512;
+export interface InternalTraceState {
+  [key: string]: string;
+}
+
+export class TraceState implements types.TraceState {
+  private internalState: InternalTraceState = {};
+
+  constructor(internalState?: InternalTraceState) {
+    if (internalState) this.internalState = internalState;
+  }
+
+  get(name: string): string | undefined {
+    return this.internalState[name];
+  }
+
+  keys(): string[] {
+    return Object.keys(this.internalState);
+  }
+
+  set(name: string, value: string): void {
+    this.internalState = {
+      // ensure that the new key ends up in the beginning of the list
+      [name]: value,
+      ...this.internalState,
+      // ensure that updates work
+      [name]: value,
+    };
+  }
+}
+
+export function serialize(state: TraceState): string {
+  return state
+    .keys()
+    .reduce((agg: string[], key) => {
+      agg.push(`${key}=${state.get(key)}`);
+      return agg;
+    }, [])
+    .join(',');
```

Comment:
This format is very difficult to work with, especially for trace states with nested objects. Why not simply rely on JSON?

Reply:
Looks like this is enforced by W3C TraceContext. We should investigate how to properly support the `,` and `=` characters, and also nested objects.

Reply:
```
Value is opaque string up to 256 characters printable ASCII [[!RFC0020]] characters (i.e., the range 0x20 to 0x7E) except comma , and =. Note that this also excludes tabs, newlines, carriage returns, etc.
```

What will go into value needs to be specified in OpenTelemetry. This is not a general purpose list for user-provided baggage.